### PR TITLE
8274179: AArch64: Support SVE operations with encodable immediates

### PR DESCRIPTION
**Pull request by fg1417**
for(int i = 0; i < LENGTH; i++) {
      c[i] = a[i] + 2;
    }

For the case showed above, after superword optimization with SVE,
without the patch, the vector add operation always has 2 z-reg inputs,
like:
mov     z16.s, #2
add	z17.s, z17.s, z16.s

Considering sve has supported basic binary operations with immediate,
this pattern could be further optimized to:
add     z16.s, z16.s, #2

To implement it, we added some new match rules and assembler rules in
the aarch64 backend. We also made some extensions on immediate types
and functions to keep backward compatible.

With the patch, only these binary integer vector operations, +(add),
-(sub), &(and), |(orr), and ^(eor) with immediate are supported for
the optimization. Other vector operations are not supported currently.

Tested tier1 and test/hotspot/jtreg/compiler on SVE featured AArch64
CPU, no new failure.

There is no obvious performance uplift but it can help remove one
redundant mov instruction.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8274179](https://bugs.openjdk.java.net/browse/JDK-8274179): AArch64: Support SVE operations with encodable immediates

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)
 * [Nick Gasson](https://openjdk.java.net/census#ngasson) (@nick-arm - **Reviewer**) ⚠️ Review applies to 05c712d4d4cbd59047f25f970b90359a9994d225

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6115/head:pull/6115` \
`$ git checkout pull/6115`

Update a local copy of the PR: \
`$ git checkout pull/6115` \
`$ git pull https://git.openjdk.java.net/jdk pull/6115/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6115`

View PR using the GUI difftool: \
`$ git pr show -t 6115`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6115.diff">https://git.openjdk.java.net/jdk/pull/6115.diff</a>

</details>

### CODE REVIEW

**Review by nick-arm — APPROVED** [review]

**Comment by fg1417** [comment]
@theRealAph , could you please help approve it? Thanks for your time :)

**Comment by fg1417** [comment]
Hi @theRealAph , I rebased my patch and retested it internally. Can I have your review :)? Thanks.

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 1]
```diff
@@ -2685,12 +2685,55 @@ bool size_fits_all_mem_uses(AddPNode* addp, int shift) {
...
+
 // Should the matcher clone input 'm' of node 'n'?
 bool Matcher::pd_clone_node(Node* n, Node* m, Matcher::MStack& mstack) {
   // ShiftV src (ShiftCntV con)
   // StoreVector (VectorStoreMask src)
+  // Binary src (Replicate con)
   if (is_vshift_con_pattern(n, m) ||
-      (UseSVE > 0 && m->Opcode() == Op_VectorStoreMask && n->Opcode() == Op_StoreVector)) {
+      (UseSVE > 0 && m->Opcode() == Op_VectorStoreMask && n->Opcode() == Op_StoreVector) ||
+      is_vector_arith_imm_pattern(n, m)) {
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64.ad:2736]: Indent this line.

**Review by theRealAph — APPROVED** [review]
Good job, well done.

**Comment by fg1417** [comment]
Thanks :) @theRealAph

**Pull request closed without merging**